### ISSUE TITLE
Website posts not federating to Mastodon

### ISSUE THREAD

**Issue by jasonrbriggs** [snarfed/bridgy-fed#56]
The first post I attempted to federate appeared on Mastodon, but the second one didn't.  Not sure i get why.

I got a 202 response:
```
$ curl -X POST -si https://fed.brid.gy/webmention -d source=http://jasonrbriggs.com/micro/2019/08/12/195738.html -d target=https://fed.brid.gy
HTTP/2 202 
content-type: text/html; charset=utf-8
cache-control: no-cache
x-cloud-trace-context: b745c55391414814571a89ec4bf76419
date: Mon, 19 Aug 2019 06:56:55 GMT
server: Google Frontend
content-length: 0
```

I don't see anything spectacularly interesting in the log...

https://fed.brid.gy/log?start_time=1566197814&key=http%3A%2F%2Fjasonrbriggs.com%2Fmicro%2F2019%2F08%2F12%2F195738.html+https%3A%2F%2Fmastodon.social%2Finbox

...other than i'm not sure about the url here:
```
2019-08-19 06:56:54.512073 I requests.post https://mastodon.social/inbox {'headers': {'Date': u'...', 'Content-Type': u'...', u'User-Agent': u'...'}, 'json': {u'@context': u'https://www.w3.org/ns/activitystreams', u'object': {u'url': u'https://fed.brid.gy/r/https://fed.brid.gy/',
```

i think the post on my site looks okay: http://jasonrbriggs.com/micro/2019/08/12/195738.html.  I'm using granary.io for my atom feed and I can see the post there as well:
https://granary.io/url?input=jsonfeed&output=atom&url=http%3A//jasonrbriggs.com/micro/feed.json
The one difference I see there is that the id in the atom feed is "https://" versus what I original sent in the webmention to fed.brid.gy (http) -- however I resubmitted using https and that didn't seem to make any difference (same 202 response).

Any ideas where I might be going wrong?

**Comment by snarfed** [comment]
hey again! thanks for all the details and sleuthing, and apologies for the trouble.

sadly i probably can't help much. we often get 202 (`Accepted`) responses from Mastodon, and sometimes those requests just disappear and the pats never show up. we've successfully debugged some of these before, but not all. mastodon.social is obviously the flagship instance, and on a recent software version, so i doubt it's misconfiguration on their part. I'm open to ideas!

**Comment by snarfed** [comment]
related: #46, #16

**Comment by jasonrbriggs** [comment]
Is there any way to get the raw json that fed.brid.gy is sending to mastodon?  Because I'm also federating using tinysubversions.com's RSS to ActivityPub converter (@jasonrbriggs@bots.tinysubversions.com), and that seems to be working fine.  So wondering if I can figure out how to source the json from that and compare with what fed.brid.gy is generating and see what the difference is (if there is a difference)?

**Comment by snarfed** [comment]
sure! that's a great idea. the full request json is in the second to last line of the log: https://fed.brid.gy/log?start_time=1566197814&key=http%3A%2F%2Fjasonrbriggs.com%2Fmicro%2F2019%2F08%2F12%2F195738.html+https%3A%2F%2Fmastodon.social%2Finbox

**Comment by snarfed** [comment]
(and apologies, I'd jump in myself, but I'm on vacation without a computer for the next week! 😁)

**Comment by jasonrbriggs** [comment]
I know the feeling. I was on vacation as well -- but took the laptop.  Kudos for leaving yours behind.

Anyway, after looking at the rss->activitypub converter I couldn't immediately see what it's generating because there's no outbox property on the user profile.  I didn't go digging into the code, because looking at the brid.gy log, I think I might see what the problem is.  If you look at this formatted json (I've removed the guts of the content to shorten it)...

```json
{
  "@context": "https://www.w3.org/ns/activitystreams",
  "object": {
    "url": "https://fed.brid.gy/r/https://fed.brid.gy/",
    "attributedTo": [
      {
        "type": "Person",
        "name": "jasonrbriggs"
      }
    ],
    "content": "<p></p>",
    "cc": [
      "https://www.w3.org/ns/activitystreams#Public"
    ],
    "published": "2019-08-12T19:57:38+07:00",
    "@context": "https://www.w3.org/ns/activitystreams",
    "type": "Note",
    "id": "https://fed.brid.gy/r/https://fed.brid.gy/"
  },
  "type": "Update"
}
```

...I think the issue is either the ID prop, the URL prop or both.  More likely to be the ID.  According to the activitystreams spec, the id:

> Provides the globally unique identifier for an Object or Link.  

The above id doesn't look like it's going to be globally unique.  That would explain why my first post appeared in mastodon, and subsequent posts are not - because mastodon won't accept something that's (from it's perspective) already there. I think the url might need to point back at the source post as well, since the spec says:

> Identifies one or more links to representations of the object 

Hope you enjoy your holiday, see what you think when you get back...

Enjoy your holiday

**Comment by snarfed** [comment]
ah! good eye, that may indeed be the problem.

good news is, you can actually fix this yourself. the url and id values come from your own post's `u-url`. you've accidentally set it to https://fed.brid.gy/ - see http://pin13.net/mf2/?url=http%3A%2F%2Fjasonrbriggs.com%2Fmicro%2F2019%2F08%2F12%2F195738.html - but it should actually be your post's canonical permalink, eg http://jasonrbriggs.com/micro/2019/08/12/195738.html . background in http://microformats.org/wiki/h-entry#Core_Properties . or you can just leave it blank.

of course, bridgy fed should have noticed this and complained in the first place. I'll make it do that too.

**Comment by snarfed** [comment]
ah. you didn't actually set this u-url yourself, it got implied. ugh. http://microformats.org/wiki/microformats2-parsing#parsing_for_implied_properties

not my favorite mf2 feature. i've mentioned this in https://chat.indieweb.org/microformats/2019-08-24#t1566658790248200

**Comment by jasonrbriggs** [comment]
So is fed.brid.gy looking at the html or looking at the json?  I've got both set in the feed.json, but I don't have u-url in the html. If adding it will fix I can easily do that...

**Comment by snarfed** [comment]
it parses your html. (pretty much all mf2 consumers do; i don't really know any that look for or use an explicit separate mf2 json feed. very unusual.) and yes, adding u-url to your post(s) should fix this.

**Comment by jasonrbriggs** [comment]
Id's and url's look a lot better now:

```json
{
  "@context": "https://www.w3.org/ns/activitystreams",
  "object": {
    "url": "https://fed.brid.gy/r/https://jasonrbriggs.com/micro/2019/08/12/195738.html",
    "attributedTo": [
      {
        "type": "Person",
        "name": "jasonrbriggs"
      }
    ],
    "content": "<p>F</p>",
    "cc": [
      "https://www.w3.org/ns/activitystreams#Public"
    ],
    "published": "2019-08-12T19:57:38+07:00",
    "@context": "https://www.w3.org/ns/activitystreams",
    "type": "Note",
    "id": "https://fed.brid.gy/r/https://jasonrbriggs.com/micro/2019/08/12/195738.html"
  }
}
```

**Comment by jasonrbriggs** [comment]
Doesn't seem to have worked. Dang. Thought that was going to be it.

**Comment by snarfed** [comment]
ugh. damn. thanks for your continued sleuthing!

**Comment by jasonrbriggs** [comment]
Woot! It's fixed.  I realised looking at the logs again that the fed.brid.gy AP request was...

```
"type": "Update"
```

...probably because I'd sent the POST request for those before and I guess that fed.brid.gy has recorded that (and mastodon is perhaps ignoring the "update".  As soon as I submitted a new post (http://jasonrbriggs.com/micro/2019/08/25/172853.html) I can see...

```
"type": "Create"
```

And the post instantly appears in mastodon.
